model/src/order.rs

```diff
@@ -42,7 +40,9 @@ impl Order {
         order_creation: OrderCreation,
         domain: &DomainSeparator,
     ) -> Option<Self> {
-        let owner = order_creation.validate_signature(domain)?;
+        let owner = order_creation
```

Comment:
I'd slightly prefer if order_creation would still expose valid_signature which does this under the hood (so that this caller here doesn't have to know too much about the internals of the struct). This way the trait would also become valuable as `validate_signature(domain)` could be implemented on the trait (using digest) and only `digest` has to be implemented by OrderCreation and OrderSignature.

Reply:
Yes, this makes perfect sense, my approach here will be to add this snippet to both `OrderCancellation` and `OrderCreation` 

```rust
    pub fn validate_signature(&self, domain_separator: &DomainSeparator) -> Option<H160> {
        self.signature.validate(domain_separator, &self.digest())
    }
```

Which would result in minimal code duplication, but I would like to take this one step further so that the snippet only has to be written once. My only issue is that this method requires access to the signature field of each struct which I can't implement on the trait level... Am I missing something or is this really as generic as it can get?

Reply:
This is the commit making the suggested change as proposed with the snippet. Just want to confirm that this subtle code duplication can't avoided:

https://github.com/gnosis/gp-v2-services/pull/331/commits/40005030af34371ac8e12c92bd8baf5edb694be3

Reply:
If the trait exposes a digest() method, you should be able to put your snippet from above as a default implementation into the trait directly (instead of implementing it on both OrderCancellation and OrderCreation)

Reply:
I will try this out, but I was more concerned about the `struct.signature` field that the method accesses.

Reply:
I believe the trait would also require a signature() method, which could be implemented by returning `self.signature`.

Reply:
Yes, this would do it! I'll give that a try.